plonky2/src/hash/field_merkle_tree.rs

```diff
@@ -0,0 +1,322 @@
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+
+use crate::hash::hash_types::RichField;
+use crate::hash::merkle_proofs::MerkleProof;
+use crate::hash::merkle_tree::{
+    capacity_up_to_mut, fill_digests_buf, merkle_tree_prove, MerkleCap,
+};
+use crate::plonk::config::{GenericHashOut, Hasher};
+use crate::util::log2_strict;
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FieldMerkleTree<F: RichField, H: Hasher<F>> {
+    /// The data stored in the Merkle tree leaves.
+    pub leaves: Vec<Vec<Vec<F>>>,
+
+    /// Merkle tree node hashes, analogous to `digests` in `MerkleTree`.
+    pub digests: Vec<H::Hash>,
+
+    /// Represents the roots of the Merkle tree. This allows for using any layer as the root of the tree.
+    pub cap: MerkleCap<F, H>,
+
+    /// Represents the heights at which leaves reside within the tree.
+    pub leaf_heights: Vec<usize>,
+}
+
+impl<F: RichField, H: Hasher<F>> Default for FieldMerkleTree<F, H> {
+    fn default() -> Self {
+        Self {
+            leaves: Vec::new(),
+            digests: Vec::new(),
+            cap: MerkleCap::default(),
+            leaf_heights: Vec::new(),
+        }
+    }
+}
+
+impl<F: RichField, H: Hasher<F>> FieldMerkleTree<F, H> {
+    /// `leaves` is a matrix (vector of vectors).
+    /// All `leaves` should have a power of two height.
+    /// All `leaves` should have different heights.
+    /// The vector of `leaves` should be sorted by height, from tallest to shortest.
+    pub fn new(mut leaves: Vec<Vec<Vec<F>>>, cap_height: usize) -> Self {
+        assert!(leaves.iter().all(|leaf| leaf.len().is_power_of_two()));
+        assert!(leaves
+            .windows(2)
+            .all(|pair| { pair[0].len() > pair[1].len() }));
+
+        let leaves_len = leaves[0].len();
```

Comment:
Here you require that leaves ain't empty, but you did not describe that in the doc comment.

Could you either make empty leaves work, or describe the limitation in the doc comments?  (And perhaps add a proper assert for it, instead of an indirect one?)

Reply:
Added the assert